lib/benchee/statistics/percentile.ex

```diff
@@ -0,0 +1,86 @@
+defmodule Benchee.Statistics.Percentile do
+  @moduledoc false
+
+  @doc """
+  Calculates the value at the `percentile_number`-th percentile. Think of this as the
+  value below which `percentile_number` percent of the samples lie. For example,
+  if `Benchee.Statistics.Percentile.percentile(samples, 99)` == 123.45,
+  99% of samples are less than 123.45.
+
+  ## Examples
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, 100)
+  5.0
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, 150)
+  5.0
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, 0)
+  1.0
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, -1)
+  1.0
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 50)
+  3.0
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 75)
+  4.75
+  """
+  @spec percentile(list(number()), integer()) :: float()
+  def percentile(samples, percentile_number) do
+    percentile(samples, length(samples), percentile_number)
+  end
+
+  @spec percentile(list(number()), integer(), integer()) :: float()
+  def percentile(samples, number_of_samples, percentile_number) when percentile_number > 100 do
```

Comment:
First off, do we want `percentile/3` to be public? It seems to me like the caller can in all cases use `percentile/2` and be totally cool.

What would you think about failing in this case and the one on line 40, and giving a helpful error message? I can't imagine that a user would give us a percentage of more than 99 or less than 1 on purpose. I think it would be likely that these would be cases of typos, and that by coercing these error cases we might be giving users results they think are strange and hiding their typos when we should be telling them how to fix their typos and get the results they want.

Reply:
@devonestes so....the reason I have `percentile/3` is because `Statistics` has both `samples` and `iterations`. I assumed that was to avoid calculating the length of the list over and over again (though I'm not certain about that). So—you're right, `percentile/2` is really the only _necessary_ interface, unless we want to be able to pass in `number_of_samples` (i.e. `iterations`) for efficiency. Have we.....benchmarked? 😁

Reply:
> What would you think about failing in this case and the one on line 40, and giving a helpful error message?

Actually that was my first inclination, but then I backed it out. What do you imagine, throwing something like `raise ArgumentError, "bad percentile value: #{bad_value}. Percentile must be greater than 0 and less than 100"`?

Reply:
For performance see #139 - with what is reasonable to expect to hit us (~1 Million) it takes ~4ms on my machine, we might have multiple scenarios so let's say * 10 that gives 40ms which is ok I guess ;) For 10 Million we'd be at about 300ms total according to this calculation. Non noticable for the given scenario user wise.
However, that's not the real reason I reused `iterations` - it just feels bad to me when I have calculated a value in a top level method and then just recalculate the value in a method directly called from that method. Such waste, could just pass it along.
When I eventually extract a library called `statistex` from benchee I think I'd keep the `/3` version because I'd like to give that possibility to people as they might operate on bigger data and it might make a different for them.

Raising an argument error sounds nice! :+1:

Reply:
so, not sure where we stand with `percentile/2` vs `percentile/3`. Speed-wise, not a big deal (thanks @PragTob). Actually, `percentile/2` is not used here at all, I just included it because it felt like the best interface. And I think there is a difference between carrying along the precomputed length within a module, versus allowing it across modules. I mean, if you pass the wrong number there, you get wrong results. So, for that reason, I'm going to remove the `percentile/3` version. It's trivial to add back again in the future.